creusot/src/backend/program.rs

```diff
@@ -145,62 +145,85 @@ pub(crate) fn translate_function<'tcx, 'sess>(
     let tcx = ctx.tcx;
     let mut names = Dependencies::new(tcx, [def_id]);
 
-    let Some(body_ids) = collect_body_ids(ctx, def_id) else {
+    let Some((body_id, promoteds)) = collect_body_ids(ctx, def_id) else {
         let (_, clones) = names.provide_deps(ctx, GraphDepth::Deep);
         return (clones, None);
     };
-    let body = to_why(ctx, &mut names, body_ids[0]);
+    let body = Decl::Coma(to_why(ctx, &mut names, body_id));
 
     if ctx.tcx.is_closure_like(def_id) {
         closure_aux_defs(ctx, def_id)
     };
 
-    let promoteds =
-        body_ids[1..].iter().map(|body_id| to_why(ctx, &mut names, *body_id)).collect::<Vec<_>>();
+    let promoteds = promoteds
+        .iter()
+        .map(|body_id| Decl::Coma(to_why(ctx, &mut names, *body_id)))
+        .collect::<Vec<_>>();
+
+    let tcx = ctx.tcx;
+    let ghost_closure_ids: std::collections::HashSet<_> = names.find_ghost_closures(tcx).collect();
+    let ghost_closures = ghost_closure_ids
+        .into_iter()
+        .map(|(def_id, subst)| {
+            // We compile the closure normally
+            closure_aux_defs(ctx, def_id);
+            let body_id = BodyId { def_id: def_id.expect_local(), promoted: None };
+            let mut coma_defn = to_why(ctx, &mut names, body_id);
+
+            // hack: to avoid having a second copy of the closure
+            names.hidden.insert(Dependency::new(tcx, (def_id, subst)));
+            // get the right name for the generated function
+            coma_defn.name = names.insert(Dependency::new(tcx, (def_id, subst))).ident();
+            Decl::Coma(coma_defn)
+        })
+        .collect::<Vec<_>>();
 
```

Comment:
This code needs to be cleaned up.

Reply:
That `hidden` hack seemed weird to me, does it really need to exists or is there another way to avoid duplication of the closure definition ?